### PR TITLE
Shared memory

### PR DESCRIPTION
**Pull request by Elon-Abulafia**
Added the ability to pass frames between components via the ram (Shared memory) instead of sending the frames through redis.

Components don't need to interact directly with the shared memory except for the first component in the pipe, meaning after the first component started using shared memory instead of redis the rest of the components will automatically interact with the shared memory via the message object.

Regarding the new classes in `shared_memory.py` - 

- `MemoryIdGenerator`: Generates a new id for a unique shared memory each time get_next is called. The id's are in the following format: `"{component_name}_{serial_memory_number}"`, this helps make a unique shared memory each time even with threads so that we won't override any existing shared memories that were already created.

- `SharedMemory`: helps simply the interaction with `posix_ipc` with 5 simple functions:
1. `release_semaphore` - Does exactly what it says...
2. `acquire_semaphore` - Does exactly what it says...
3. `write_to_memory` - writes the frame given to it to the shared memory object.
4. `read_from_memory` - reads what is currently in the shared memory.
5. `free_memory` - cleans what is on the memory and deletes it.
You can access this class via the `get_shared_memory_object` by giving it just the name of the wanted shared memory.

- `SharedMemoryGenerator`: Generates a new shared memory each time `get_next_shared_memory` is called and is responsible for cleaning up shared memories if the count that exists now exceeds the max or the proccess has ended.
(The `cleanup` method is not called automatically and should be called in the cleanup of a routine/component)

### CODE REVIEW

File: tests/pipert/core/test_shared_memory.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+
+def test_max_count():
+    generator = DummySharedMemoryGenerator()
+    first_memory = generator.get_next_shared_memory()
+    generator.get_next_shared_memory()
+    generator.get_next_shared_memory()
+    generator.get_next_shared_memory()
+    generator.get_next_shared_memory()
+    generator.get_next_shared_memory()
```
Comment (ItamarWilf) [inline tests/pipert/core/test_shared_memory.py]: plz use a for loop :)

File: pipert/core/shared_memory.py [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        """
+        next_name = "{0}_{1}".format(self.component_name, self.name_count)
+        name_to_unlink = ""
+        self.name_count += 1
+        if self.name_count >= self.max_count:
+            name_to_unlink = "{0}_{1}".format(self.component_name,
+                                              (self.name_count
+                                               - self.max_count))
+
+        return next_name, name_to_unlink
```
Comment (ItamarWilf) [inline pipert/core/shared_memory.py]: @Elon-Abulafia could you help me understand what is the difference between `next_name` and `name_to_unlink`? 
thanks

File: pipert/core/shared_memory.py [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+        memory.close_fd()
+
+        self.shared_memories[next_name] = SharedMemory(memory, semaphore,
+                                                       mapfile)
+
+        if name_to_unlink:
+            if name_to_unlink in self.shared_memories:
+                self.shared_memories[name_to_unlink].free_memory()
+                self.shared_memories.pop(name_to_unlink)
```
Comment (ItamarWilf) [inline pipert/core/shared_memory.py]: This is the same as lines 128-130, maybe it could be moved to a private method?

File: pipert/core/shared_memory.py (hunk 2 above)
Comment (ItayHoresh) [reply pipert/core/shared_memory.py]: @ItamarWilf every routine has a max size of shared memory objects in the same time.
if the object name_to_unlink is still alive when we want to allocate a new shared memory object next_name, we release the name_to_unlink object and replace it with next_name

**Comment by Elon-Abulafia** [comment]
@ItamarWilf Regarding your general questions:
1. What part exactly are you referring to here exactly so I can answer accordingly?
2. I think that because the way we communicate with redis doesn't really change when using the shared memory, there isn't really any need for a new message handler, but instead we just need a new payload type that is hidden and handles all of the interactions with the shared memory.
Also because every routine that cares about the way they save data into the shared memory will use it in a different way, I think we need to enable them to interact with it how they need to instead of forcing them to use a prebuilt message handler. (I might be wrong here and if I am I can try and write a new message handler, maybe one that'll expand on `RedisHandler`)

**Comment by ItamarWilf** [comment]
We should add:

- [ ] a sequence diagram explaining the flow of data and management
- [ ] required tests for the memory management and edge cases
